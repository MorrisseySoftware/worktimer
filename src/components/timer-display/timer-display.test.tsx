import React from 'react'
import { act, create, ReactTestRenderer } from 'react-test-renderer'
import { shallow, mount, render } from 'enzyme'
import TimerDisplay, { TimeState } from './timer-display'
import toJson from 'enzyme-to-json'
import {
    mockDate_Add10Mins,
    mockDate_Add1Mins,
    mockDate_Static_Past,
} from '../../dev/test.utls'
import { printClockTime, printRemainingTime } from '../../utilities/time.utils'

describe('-Timer Display Tests-', () => {
    beforeEach(() => {
        mockDate_Static_Past()
    })
    describe('-Prop Tests-', () => {
        it('-should be hidden when run is false-', () => {
            let component = mount(
                <TimerDisplay
                    run={false}
                    completionTime={new Date(Date.now())}
                    timerCallback={() => {}}
                    completeCallback={() => {}}
                />
            )
            expect(toJson(component)).toMatchSnapshot()
        })
        it('-should be displayed when run is true-', () => {
            let component = mount(
                <TimerDisplay
                    run={true}
                    completionTime={new Date(Date.now())}
                    timerCallback={() => {}}
                    completeCallback={() => {}}
                />
            )
            expect(toJson(component)).toMatchSnapshot()
        })
        it('-should not be hidden after its been displayed-', () => {
            let component = mount(
                <TimerDisplay
                    run={true}
                    completionTime={new Date(Date.now())}
                    timerCallback={() => {}}
                    completeCallback={() => {}}
                />
            )
            component.setProps({ run: false })
            expect(toJson(component)).toMatchSnapshot()
        })
        it('-should trigger timerCallback when run value changes from true to false-', () => {
            mockDate_Add10Mins()
            let result = 'failed'
            const callbk = (val: TimeState) => {
                result = printRemainingTime(val.timeleft)
            }
            const component = mount(
                <TimerDisplay
                    run={true}
                    completionTime={new Date(Date.now())}
                    timerCallback={callbk}
                    completeCallback={() => {}}
                />
            )
            component.setProps({ run: false })
            expect(result).toBe('00:00:00')
        })
        it('-should not trigger timerCallback when run value changes from false to true-', () => {
            mockDate_Add1Mins()
            let result = 'passed'
            const callbk = (val: TimeState) => {
                result = printRemainingTime(val.timeleft)
            }
            const component = mount(
                <TimerDisplay
                    run={false}
                    completionTime={new Date(Date.now())}
                    timerCallback={callbk}
                    completeCallback={() => {}}
                />
            )
            component.setProps({ run: true })
            expect(result).toBe('passed')
        })
        it('-should trigger completeCallback when completionTime has passed-', () => {
            let result = 'failed'
            const callbk = (val: TimeState) => {
                result = printRemainingTime(val.timeleft)
            }
            const component = mount(
                <TimerDisplay
                    run={true}
                    completionTime={new Date(Date.now())}
                    timerCallback={() => {}}
                    completeCallback={callbk}
                />
            )
            expect(result).toBe('00:00:00')
        })
    })
    describe('Timer Styling Tests', () => {
        it('should be render the display green when the timer is greater than 5 mins', () => {
            mockDate_Add10Mins()
            const component = mount(
                <TimerDisplay
                    run={true}
                    completionTime={new Date(Date.now())}
                    timerCallback={() => {}}
                    completeCallback={() => {}}
                />
            )
            const container = component.find('.container')
            expect(container).toBeDefined()
            expect(container.hasClass('completed')).toBeFalsy()
            expect(container.hasClass('warning')).toBeFalsy()
        })
        it('should be render the display yellow when 0 < timer < 5 mins', () => {
            mockDate_Add1Mins()
            const component = mount(
                <TimerDisplay
                    run={true}
                    completionTime={new Date(Date.now())}
                    timerCallback={() => {}}
                    completeCallback={() => {}}
                />
            )
            const container = component.find('.container')
            expect(container).toBeDefined()
            expect(container.hasClass('completed')).toBeFalsy()
            expect(container.hasClass('warning')).toBeTruthy()
        })
        it('should be render the display red when timer = 0 mins', () => {
            mockDate_Static_Past()
            const component = mount(
                <TimerDisplay
                    run={true}
                    completionTime={new Date(Date.now())}
                    timerCallback={() => {}}
                    completeCallback={() => {}}
                />
            )
            const container = component.find('.container')
            expect(container).toBeDefined()
            expect(container.hasClass('completed')).toBeTruthy()
            expect(container.hasClass('warning')).toBeFalsy()
        })
        it('should be render the display red when timer stopped', () => {
            mockDate_Add10Mins()
            const component = mount(
                <TimerDisplay
                    run={true}
                    completionTime={new Date(Date.now())}
                    timerCallback={() => {}}
                    completeCallback={() => {}}
                />
            )
            component.setProps({ run: false })
            const container = component.find('.container')
            expect(container).toBeDefined()
            expect(container.hasClass('completed')).toBeTruthy()
            expect(container.hasClass('warning')).toBeFalsy()
        })
    })
})
