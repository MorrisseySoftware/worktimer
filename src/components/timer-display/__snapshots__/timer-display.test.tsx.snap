// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`-Timer Display Tests- -Prop Tests- -should be displayed when run is true- 1`] = `
<TimerDisplay
  completeCallback={[Function]}
  completionTime={2019-05-14T11:01:58.135Z}
  run={true}
  timerCallback={[Function]}
>
  <div
    className="container completed"
  >
    <div
      className="timer__container--base"
    >
      <p
        className="timer__label"
      >
        Time remaining
      </p>
      <p
        className="timer"
      >
        00:00:00
      </p>
      <p
        className="completion__label"
      >
        Completion Time:
         
        09:01 PM
      </p>
    </div>
  </div>
</TimerDisplay>
`;

exports[`-Timer Display Tests- -Prop Tests- -should be hidden when run is false- 1`] = `
<TimerDisplay
  completeCallback={[Function]}
  completionTime={2019-05-14T11:01:58.135Z}
  run={false}
  timerCallback={[Function]}
>
  <div
    className="container completed"
  />
</TimerDisplay>
`;

exports[`-Timer Display Tests- -Prop Tests- -should not be hidden after its been displayed- 1`] = `
<TimerDisplay
  completeCallback={[Function]}
  completionTime={2019-05-14T11:01:58.135Z}
  run={false}
  timerCallback={[Function]}
>
  <div
    className="container completed"
  >
    <div
      className="timer__container--base"
    >
      <p
        className="timer__label"
      >
        Time remaining
      </p>
      <p
        className="timer"
      >
        00:00:00
      </p>
      <p
        className="completion__label"
      >
        Completion Time:
         
        09:01 PM
      </p>
    </div>
  </div>
</TimerDisplay>
`;
